crates/proof-of-sql/src/base/database/slice_operation.rs

```diff
@@ -264,6 +265,22 @@ pub(super) fn apply_slice_to_indexes<S: Clone>(
         .collect()
 }
 
+/// Take set union of two slices and order the result.
+pub(super) fn ordered_set_union<S: Ord + Clone + Hash>(lhs: &[S], rhs: &[S]) -> Vec<S> {
+    lhs.iter().chain(rhs).unique().cloned().sorted().collect()
+}
+
+/// Get multiplicities of elements of `rhs` in `lhs`.
+///
+/// Normally all elements of the former slice should be in the latter slice
+/// and that elements in the latter slice should be unique. However, this
+/// function does not check for these conditions.
+pub(super) fn get_multiplicities<S: PartialEq>(lhs: &[S], rhs: &[S]) -> Vec<u64> {
```

Comment:
Can you add some tracing for this. I'm fine with the code for now, but we can eventually optimize it if needed because we will know that `rhs` is sorted.
```suggestion
#[tracing::instrument(name = "slice_operation::get_multiplicities", level = "debug", skip_all)]
pub(super) fn get_multiplicities<S: PartialEq>(lhs: &[S], rhs: &[S]) -> Vec<u64> {
```